packages/fiber/src/native/Canvas.tsx

```diff
@@ -0,0 +1,108 @@
+  children: React.ReactNode
+}
+
+// React currently throws a warning when using useLayoutEffect on the server.
+// To get around it, we can conditionally useEffect on the server (no-op) and
+// useLayoutEffect in the browser.
+const useIsomorphicLayoutEffect = typeof window !== 'undefined' ? React.useLayoutEffect : React.useEffect
+
+function Block({ set }: Omit<UnblockProps, 'children'>) {
+  useIsomorphicLayoutEffect(() => {
+    set(new Promise(() => null))
+    return () => set(false)
+  }, [])
+  return null
+}
+
+class ErrorBoundary extends React.Component<{ set: React.Dispatch<any> }, { error: boolean }> {
+  state = { error: false }
+  static getDerivedStateFromError = () => ({ error: true })
+  componentDidCatch(error: any) {
+    this.props.set(error)
+  }
+  render() {
+    return this.state.error ? null : this.props.children
+  }
+}
+
+export const Canvas = React.forwardRef<View, Props>(
+  ({ children, fallback, style, events, nativeRef_EXPERIMENTAL, onContextCreate, ...props }, forwardedRef) => {
+    const [context, setContext] = React.useState<GLContext | null>(null)
+    const [size, setSize] = React.useState({ width: 0, height: 0 })
+    const [bind, setBind] = React.useState()
+    const [block, setBlock] = React.useState<SetBlock>(false)
+    const [error, setError] = React.useState<any>(false)
+
+    // Suspend this component if block is a promise (2nd run)
+    if (block) throw block
+    // Throw exception outwards if anything within canvas throws
+    if (error) throw error
+
+    const onLayout = React.useCallback((e: LayoutChangeEvent) => {
+      const { width, height } = e.nativeEvent.layout
+      setSize({ width, height })
+    }, [])
+
+    // Execute JSX in the reconciler as a layout-effect
+    useIsomorphicLayoutEffect(() => {
+      if (context) {
+        const store = render(
+          <ErrorBoundary set={setError}>
+            <React.Suspense fallback={<Block set={setBlock} />}>{children}</React.Suspense>
+          </ErrorBoundary>,
+          context,
+          { ...props, size, events: events || createTouchEvents },
+        )
+
+        const state = store.getState()
+        setBind(state.events.connected.getEventHandlers())
+      }
+    }, [size, children, context])
+
+    useIsomorphicLayoutEffect(() => {
+      return () => {
+        if (context) unmountComponentAtNode(context)
+      }
+    }, [])
+
+    return (
+      <View ref={forwardedRef} onLayout={onLayout} style={{ flex: 1, ...style }} {...bind}>
+        {size.width > 0 && (
+          <GLView
+            nativeRef_EXPERIMENTAL={(ref: any) => {
+              if (nativeRef_EXPERIMENTAL && !nativeRef_EXPERIMENTAL.current) {
+                nativeRef_EXPERIMENTAL.current = ref
+              }
+            }}
+            onContextCreate={async (gl) => {
+              await onContextCreate?.(gl)
+              setContext(gl)
+            }}
+            style={StyleSheet.absoluteFill}
+          />
+        )}
+      </View>
+    )
```

Comment:
This needs to be updated with master, but we should/can be able to pass non-R3F props here, so we'll have to handle conflicts with event callbacks.